crates/hir_ty/src/infer/closure.rs

```diff
@@ -0,0 +1,89 @@
+//! Inference of closure parameter types based on the closure's expected type.
+
+use chalk_ir::{AliasTy, WhereClause};
+use hir_def::HasModule;
+use smallvec::SmallVec;
+
+use crate::{to_chalk_trait_id, utils, CallableSig, ChalkTraitId, DynTy, Interner, Ty, TyKind};
+
+use super::{Expectation, InferenceContext};
+
+impl InferenceContext<'_> {
+    pub(super) fn deduce_closure_type_from_expectations(
+        &mut self,
+        closure_ty: &Ty,
+        args_and_ret_tys: &[Ty],
+        expectation: &Expectation,
+    ) {
+        let expected_ty = match expectation.to_option(&mut self.table) {
+            Some(ty) => ty,
+            None => return,
+        };
+
+        // Deduction from where-clauses in scope, as well as fn-pointer coercion are handled here.
+        self.coerce(closure_ty, &expected_ty);
+
+        // Deduction based on the expected `dyn Fn` is done separately.
+        if let TyKind::Dyn(dyn_ty) = expected_ty.kind(&Interner) {
+            if let Some(sig) = self.deduce_sig_from_dyn_ty(dyn_ty) {
+                for (expected, arg) in sig.params().iter().zip(args_and_ret_tys) {
+                    self.unify(arg, expected);
+                }
+                let ret = args_and_ret_tys.last().unwrap();
+                self.unify(ret, sig.ret());
+            }
+        }
+    }
+
+    fn deduce_sig_from_dyn_ty(&self, dyn_ty: &DynTy) -> Option<CallableSig> {
+        // Search for predicates like `$self: FnX<Args>` and `<$self as FnOnce<...>>::Output == Ret`
+
+        let fn_traits: SmallVec<[ChalkTraitId; 3]> =
+            utils::fn_traits(self.db.upcast(), self.owner.module(self.db.upcast()).krate())
+                .map(|tid| to_chalk_trait_id(tid))
+                .collect();
+
+        let mut ret_ty = None;
+        let mut params = None;
+        for bound in dyn_ty.bounds.skip_binders().as_slice(&Interner) {
+            match bound.skip_binders() {
+                WhereClause::Implemented(trait_ref) => {
+                    if !fn_traits.contains(&trait_ref.trait_id) {
+                        continue;
+                    }
+
+                    // Skip `Self`, get the type argument.
+                    let arg = trait_ref.substitution.as_slice(&Interner).get(1)?;
+                    match arg.ty(&Interner)?.kind(&Interner) {
+                        TyKind::Tuple(_, subst) => {
+                            let generic_args = subst.as_slice(&Interner);
+                            let mut closure_params = Vec::new();
+                            for arg in generic_args {
+                                closure_params.push(arg.ty(&Interner)?.clone());
+                            }
+                            params = Some(closure_params);
```

Comment:
here where we're extracting from inside the binders, we should use `shifted_out_to` to shift any bound variables inside this (there probably can't be any? but I'd still do it on principle)